### PR TITLE
SDK-385: OpenSSL config header issues in Android builds

### PR DESCRIPTION
**Pull request by solomonchild**

### CODE REVIEW

File: android/buildinf.h [hunk 1]
```diff
@@ -0,0 +1,143 @@
+#if !__ANDROID__
+    #error "This header file should only be used for building Android native library."
+#endif
+
+#ifdef __ARM_ARCH
+    #if __ARM_ARCH_ISA_A64
```
Comment (arvids-kokins-bidstack) [inline android/buildinf.h:6]: these don't look like compiler-provided defines, where are they from?
please use the ones both GCC and Clang provide natively (e.g. `__arm__`, `__aarch64__`)

**Comment by arvids-kokins-bidstack** [comment]
could you reproduce the issue where the files aren't copied at all (when clean-rebuilding from android studio)?

File: android/buildinf.h (hunk 1 above)
Comment (solomonchild) [reply android/buildinf.h:6]: These are the ones provided by the compiler. I've got those by doing `clang++ --target=<arch>-linux-<API> -E -dM` on an empty header file.

**Comment by solomonchild** [comment]
> could you reproduce the issue where the files aren't copied at all (when clean-rebuilding from android studio)?

Yes. Both that and the warnings after manually swapping what gets `define`d and `undefine`d for `arm` architecture.

**Comment by solomonchild** [comment]
@arvids-kokins-bidstack 
BTW, I don't see neither `__arm__` nor `__aarch64__` in the outputs.
[aarch64_defs.txt](https://github.com/bidstack-group/openssl/files/4598758/aarch64_defs.txt)
[armv7a_defs.txt](https://github.com/bidstack-group/openssl/files/4598759/armv7a_defs.txt)
[i686_defs.txt](https://github.com/bidstack-group/openssl/files/4598760/i686_defs.txt)
[x86_64_defs.txt](https://github.com/bidstack-group/openssl/files/4598761/x86_64_defs.txt)

File: android/buildinf.h (hunk 1 above)
Comment (solomonchild) [reply android/buildinf.h:6]: http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0774a/chr1383660321827.html

**Comment by arvids-kokins-bidstack** [comment]
> @arvids-kokins-bidstack
> BTW, I don't see neither `__arm__` nor `__aarch64__` in the outputs.
> [aarch64_defs.txt](https://github.com/bidstack-group/openssl/files/4598758/aarch64_defs.txt)
> [armv7a_defs.txt](https://github.com/bidstack-group/openssl/files/4598759/armv7a_defs.txt)
> [i686_defs.txt](https://github.com/bidstack-group/openssl/files/4598760/i686_defs.txt)
> [x86_64_defs.txt](https://github.com/bidstack-group/openssl/files/4598761/x86_64_defs.txt)

I see them:

image

image

**Comment by solomonchild** [comment]
@arvids-kokins-bidstack right, I was probably looking for `__arm__` in arm64 file

**Comment by arvids-kokins-bidstack** [comment]
> Yes. Both that and the warnings after manually swapping what gets `define`d and `undefine`d for `arm` architecture.

ok, does this PR address that, and what was the underlying issue behind the files not being copied?

**Comment by solomonchild** [comment]
> ok, does this PR address that, and what was the underlying issue behind the files not being copied?

The thing is that Clean does not remove CMake artifacts (located in android/bidstack-sdk/.cxx). If those get removed then CMake does the generation step and the necessary file gets copied. 
So this PR addresses that indirectly: only one copy is required for all the architectures.

File: android/buildinf.h (hunk 1 above)
Comment (arvids-kokins-bidstack) [reply android/buildinf.h:6]: ok, though I still don't see `__ARM_ARCH_ISA_ARM` in the page 🤔

**Comment by arvids-kokins-bidstack** [comment]
> The thing is that Clean does not remove CMake artifacts (located in android/bidstack-sdk/.cxx). If those get removed then CMake does the generation step and the necessary file gets copied.
> So this PR addresses that indirectly: only one copy is required for all the architectures.

~and what about building windows and android from the same source?~ nvm, forgot that we don't build openssl on windows

would it work to build mac/android from the same source though?

**Comment by solomonchild** [comment]
>would it work to build mac/android from the same source though

Darwin and linux builds still generate configuration files on the fly using a Perl script. Probably need to make a future task to get rid of that and use pre-generated files for everything.

File: android/buildinf.h (hunk 1 above)
Comment (solomonchild) [reply android/buildinf.h:6]: Yeah, I only found it in this doc: https://static.docs.arm.com/ihi0053/c/IHI0053C_acle_2_0.pdf
I don't insist on having `ISA_*` TBH, we can use  `__arm__` and `__aarch_64__`.

**Comment by arvids-kokins-bidstack** [comment]
> Darwin and linux builds still generate configuration files on the fly using a Perl script. Probably need to make a future task to get rid of that and use pre-generated files for everything.

ok, thanks, will need to think about that

openssl is proving to be quite the pain compared to boringssl, but maybe there's something even better available somewhere

**Review by arvids-kokins-bidstack — APPROVED** [review]

File: android/buildinf.h (hunk 1 above)
Comment (arvids-kokins-bidstack) [reply android/buildinf.h:6]: it's ok, I just picked up some old ndk's gcc and didn't see them there, but even old versions of android clang seem to have it

**Review by a13ks — APPROVED** [review]

**Pull request merged**